### PR TITLE
OP Stack SOLC update from 0.8.15 to 0.8.25

### PR DESCRIPTION
**Pull request by AmadiMichael**
### Description

This covers solidity version upgrade of all OP Stack smart contracts from 0.8.15 code to 0.8.25.

### Metadata

Fixes https://github.com/ethereum-optimism/optimism/issues/11527

### CODE REVIEW

**Review by smartcontracts — COMMENTED** [review]
Looks good to me. I've already managed to compile the contracts with solc 0.8.25 and it worked fine. We also seem to have mercifully gained back a few hundred bytes inside of OptimismPortal2Interop.

**Review by smartcontracts — APPROVED** [review]

**Comment by smartcontracts** [comment]
You have my approval but please also get approval from @mds1 before merging

**Comment by smartcontracts** [comment]
Oh and @maurelian too

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (maurelian) [inline solidity/solc-update-0.8.15-to-0.8.25.md:58]: I think this is leftover from the template?

Comment (AmadiMichael) [reply solidity/solc-update-0.8.15-to-0.8.25.md:58]: Yeah it is, I left it because I believe it helps the reader understand what is written afterwards

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (mds1) [inline solidity/solc-update-0.8.15-to-0.8.25.md:23]:
```suggestion
| Other Reviewers     | Matt Solomon, maurelian   |
```

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (mds1) [inline solidity/solc-update-0.8.15-to-0.8.25.md:19]: If we like this table (which I do), let's also add it to the template

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (mds1) [inline solidity/solc-update-0.8.15-to-0.8.25.md:35]: leftover from template
```suggestion
```

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (mds1) [inline solidity/solc-update-0.8.15-to-0.8.25.md:38]: This is saying if I have:
```solidity
mapping (address from => mapping (address to => uint256)) public allowance;
```

I can call it like this?
```solidity
allowance({from: me, to: you})
```

That's cool, I didn't realized the named param syntax worked for mappings

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (mds1) [inline solidity/solc-update-0.8.15-to-0.8.25.md:47]: mcopy will also simplify our proofs and may improve performance there as a result

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (mds1) [inline solidity/solc-update-0.8.15-to-0.8.25.md:87]: Looks like this is from the FMA template. Do we want to add this to the solc template as well? I think it might be a good idea as it helps us sanity check we don't e.g. exceed code size limits when changing versions. Any thoughts @smartcontracts @maurelian

**Review by mds1 — APPROVED** [review]

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (AmadiMichael) [reply solidity/solc-update-0.8.15-to-0.8.25.md:38]: Wait, actually this isn't possible 🤦‍♂️, will remove it

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (AmadiMichael) [reply solidity/solc-update-0.8.15-to-0.8.25.md:19]: @smartcontracts

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (maurelian) [inline solidity/solc-update-0.8.15-to-0.8.25.md:77]: The only failure modes I see are listed below. Is that what you're referring to here?

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (maurelian) [inline solidity/solc-update-0.8.15-to-0.8.25.md:79]: Feels like boiler plate to me
```suggestion
```

**Review by maurelian — APPROVED** [review]
Left a couple suggestions, am good once resolved.

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (AmadiMichael) [reply solidity/solc-update-0.8.15-to-0.8.25.md:77]: Yeah, i will modify this now

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (AmadiMichael) [reply solidity/solc-update-0.8.15-to-0.8.25.md:79]: Do you suggest I remove it entirely or only replace the `document author` text?

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (AmadiMichael) [reply solidity/solc-update-0.8.15-to-0.8.25.md:43]: Not sure if you suggest I add this or just a comment...

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (maurelian) [reply solidity/solc-update-0.8.15-to-0.8.25.md:79]: remove entirely

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (maurelian) [inline solidity/solc-update-0.8.15-to-0.8.25.md]: why was this removed?

**Review by maurelian — APPROVED** [review]

File: solidity/solc-update-0.8.15-to-0.8.25.md
Comment (AmadiMichael) [reply solidity/solc-update-0.8.15-to-0.8.25.md]: Because its actually not possible. https://github.com/ethereum-optimism/design-docs/pull/106#discussion_r1786445376, https://github.com/ethereum-optimism/design-docs/pull/106#discussion_r1786462875

**Pull request merged**
